File: src/services/typeHierarchy.ts
// Type hierarchy (supertypes / subtypes), as pure functions over the program -
// the transport-free half, mirroring how TypeScript keeps its callHierarchy /
// typeHierarchy logic in a service module separate from the protocol layer. The
// server maps the returned LSP TypeHierarchyItems to the wire and re-enters
// `supertypes`/`subtypes` with the item the client hands back; each call
// re-resolves the type symbol from the item's selectionRange position, so no
// symbol identity has to survive the round-trip.

import { type Range, SymbolKind, type TypeHierarchyItem } from "vscode-languageserver-types";

import {
  type Character,
  computeLineStarts,
  getLineAndCharacterOfPosition,
  getPositionOfLineAndCharacter,
  type Line,
} from "../compiler/lineMap.ts";
import type { Program } from "../compiler/program.ts";
import {
  getDeclarationNameNode,
  getDirectSuperTypeSymbols,
  getSourceFileOfNode,
} from "../compiler/resolver.ts";
import { type Checker } from "../compiler/checker.ts";
import {
  type Identifier,
  type SourceFile,
  type Symbol,
  SymbolFlags,
} from "../compiler/types.ts";
import { skipTrivia } from "../compiler/utilities.ts";
import { type Uri } from "../workspace.ts";
import { getIdentifierAtPosition } from "./nodeAtPosition.ts";
import { getSubtypeIndex } from "./subtypes.ts";

const TYPE_FLAGS =
  SymbolFlags.Class |
  SymbolFlags.Interface |
  SymbolFlags.Enum |
  SymbolFlags.Record |
  SymbolFlags.Annotation;

function symbolKindOf(flags: SymbolFlags): SymbolKind {
  if (flags & (SymbolFlags.Interface | SymbolFlags.Annotation)) return SymbolKind.Interface;
  if (flags & SymbolFlags.Enum) return SymbolKind.Enum;
  return SymbolKind.Class; // class and record
}

function rangeOf(text: string, lineStarts: readonly number[], pos: number, end: number): Range {
  return {
    start: getLineAndCharacterOfPosition(lineStarts, pos),
    end: getLineAndCharacterOfPosition(lineStarts, end),
  };
}

// Build a TypeHierarchyItem for a type symbol, or undefined if it has no source
// declaration (e.g. a JDK-stub type). selectionRange is the type's name; range
// spans the whole declaration.
function itemOf(symbol: Symbol): TypeHierarchyItem | undefined {
  const nameNode = getDeclarationNameNode(symbol);
  const declaration = symbol.valueDeclaration ?? symbol.declarations?.[0];
  if (!nameNode || !declaration) return undefined;
  const file = getSourceFileOfNode(declaration);
  const lineStarts = computeLineStarts(file.text);
  return {
    name: (nameNode as { text?: string }).text ?? "<anonymous>",
    kind: symbolKindOf(symbol.flags),
    uri: file.fileName,
    range: rangeOf(file.text, lineStarts, skipTrivia(file.text, declaration.pos), declaration.end),
    selectionRange: rangeOf(file.text, lineStarts, skipTrivia(file.text, nameNode.pos), nameNode.end),
  };
}

// Resolve the type symbol an item points at, by re-resolving the identifier at
// its selectionRange start (the type's name).
function typeSymbolOfItem(program: Program, checker: Checker, item: TypeHierarchyItem): Symbol | undefined {
  const sourceFile = program.getSourceFile(item.uri as Uri);
  if (!sourceFile) return undefined;
  const lineStarts = computeLineStarts(sourceFile.text);
  const offset = getPositionOfLineAndCharacter(
    lineStarts,
    item.selectionRange.start.line as Line,
    item.selectionRange.start.character as Character,
  );
  const id = getIdentifierAtPosition(sourceFile, offset) as Identifier | undefined;
  const symbol = id && checker.resolveName(id);
  return symbol && symbol.flags & TYPE_FLAGS ? symbol : undefined;
}

// The type at a position, as a single-element item list (or null for "no type").
export function prepareTypeHierarchy(
  checker: Checker,
  sourceFile: SourceFile,
  offset: number,
): TypeHierarchyItem[] | null {
  const id = getIdentifierAtPosition(sourceFile, offset) as Identifier | undefined;
  if (!id) return null;
  const symbol = checker.resolveName(id);
  if (!symbol || !(symbol.flags & TYPE_FLAGS)) return null;
  const item = itemOf(symbol);
  return item ? [item] : null;
}

// The direct supertypes (extends + implements) of an item's type.
export function typeHierarchySupertypes(
  program: Program,
  checker: Checker,
  item: TypeHierarchyItem,
): TypeHierarchyItem[] | null {
  const symbol = typeSymbolOfItem(program, checker, item);
  if (!symbol) return null;
  const items = getDirectSuperTypeSymbols(symbol, program)
    .map(itemOf)
    .filter((i): i is TypeHierarchyItem => i !== undefined);
  return items.length > 0 ? items : null;
}

// The direct subtypes (declarations whose extends/implements names this type).
export function typeHierarchySubtypes(
  program: Program,
  checker: Checker,
  item: TypeHierarchyItem,
): TypeHierarchyItem[] | null {
  const symbol = typeSymbolOfItem(program, checker, item);
  if (!symbol) return null;
  const items = getSubtypeIndex(program)
    .directSubtypesOf(symbol)
    .map(itemOf)
    .filter((i): i is TypeHierarchyItem => i !== undefined);
  return items.length > 0 ? items : null;
}

File: togo/internal/services/type_hierarchy.go
package services

// Type hierarchy (supertypes / subtypes) as pure functions over the program -
// the transport-free half, kept separate from the lspserver layer for testing.
// Port of src/services/typeHierarchy.ts. Supertypes/subtypes re-resolve the type
// symbol from the item's SelectionRange position, so no opaque data survives the
// round-trip.

import (
	"github.com/nikeee/cappu/internal/compiler"
	"github.com/nikeee/cappu/internal/lsp"
)

func symbolKindOf(flags compiler.SymbolFlags) lsp.SymbolKind {
	switch {
	case flags&(compiler.SymbolFlagsInterface|compiler.SymbolFlagsAnnotation) != 0:
		return lsp.SymbolKindInterface
	case flags&compiler.SymbolFlagsEnum != 0:
		return lsp.SymbolKindEnum
	default:
		return lsp.SymbolKindClass // class and record
	}
}

// typeHierarchyItemOf builds an item for a type symbol; ok is false when it has
// no source declaration (e.g. a JDK-stub type).
func typeHierarchyItemOf(symbol *compiler.Symbol) (lsp.TypeHierarchyItem, bool) {
	nameNode := compiler.GetDeclarationNameNode(symbol)
	declaration := symbolDeclaration(symbol)
	if nameNode == nil || declaration == nil {
		return lsp.TypeHierarchyItem{}, false
	}
	file := compiler.GetSourceFileOfNode(declaration)
	text := file.AsSourceFile().Text
	lineStarts := file.AsSourceFile().LineStarts()
	name := "<anonymous>"
	if nameNode.Kind == compiler.Identifier {
		name = nameNode.AsIdentifier().Text
	}
	return lsp.TypeHierarchyItem{
		Name:           name,
		Kind:           symbolKindOf(symbol.Flags),
		URI:            file.AsSourceFile().FileName,
		Range:          dsRange(text, lineStarts, compiler.SkipTrivia(text, declaration.Pos), declaration.End),
		SelectionRange: dsRange(text, lineStarts, compiler.SkipTrivia(text, nameNode.Pos), nameNode.End),
	}, true
}

// typeSymbolOfItem re-resolves the type an item points at, via the identifier at
// its SelectionRange start.
func typeSymbolOfItem(program *compiler.Program, checker *compiler.Checker, item lsp.TypeHierarchyItem) *compiler.Symbol {
	sourceFile := program.GetSourceFile(compiler.URI(item.URI))
	if sourceFile == nil {
		return nil
	}
	text := sourceFile.AsSourceFile().Text
	lineStarts := sourceFile.AsSourceFile().LineStarts()
	offset := compiler.GetPositionOfLineAndCharacter(text, lineStarts, item.SelectionRange.Start.Line, item.SelectionRange.Start.Character)
	id := compiler.GetIdentifierAtPosition(sourceFile, offset)
	if id == nil {
		return nil
	}
	symbol := checker.ResolveName(id)
	if symbol == nil || symbol.Flags&typeFlags == 0 {
		return nil
	}
	return symbol
}

// PrepareTypeHierarchy returns the type at a position as a single-element list,
// or nil when the cursor is not on a type.
func PrepareTypeHierarchy(checker *compiler.Checker, sourceFile *compiler.Node, offset int) []lsp.TypeHierarchyItem {
	id := compiler.GetIdentifierAtPosition(sourceFile, offset)
	if id == nil {
		return nil
	}
	symbol := checker.ResolveName(id)
	if symbol == nil || symbol.Flags&typeFlags == 0 {
		return nil
	}
	if item, ok := typeHierarchyItemOf(symbol); ok {
		return []lsp.TypeHierarchyItem{item}
	}
	return nil
}

// TypeHierarchySupertypes returns the direct supertypes (extends + implements).
func TypeHierarchySupertypes(program *compiler.Program, checker *compiler.Checker, item lsp.TypeHierarchyItem) []lsp.TypeHierarchyItem {
	symbol := typeSymbolOfItem(program, checker, item)
	if symbol == nil {
		return nil
	}
	var out []lsp.TypeHierarchyItem
	for _, s := range compiler.GetDirectSuperTypeSymbols(symbol, program) {
		if it, ok := typeHierarchyItemOf(s); ok {
			out = append(out, it)
		}
	}
	return out
}

// TypeHierarchySubtypes returns the direct subtypes (declarations whose
// extends/implements names this type).
func TypeHierarchySubtypes(program *compiler.Program, checker *compiler.Checker, item lsp.TypeHierarchyItem) []lsp.TypeHierarchyItem {
	symbol := typeSymbolOfItem(program, checker, item)
	if symbol == nil {
		return nil
	}
	var out []lsp.TypeHierarchyItem
	for _, s := range GetSubtypeIndex(program).DirectSubtypesOf(symbol) {
		if it, ok := typeHierarchyItemOf(s); ok {
			out = append(out, it)
		}
	}
	return out
}
